rpc.go

```diff
@@ -293,11 +303,106 @@ func (l *Libvirt) deregister(id uint32) {
 // returns response returned by server.
 // if response is not OK, decodes error from it and returns it.
 func (l *Libvirt) request(proc uint32, program uint32, payload []byte) (response, error) {
+	return l.requestStream(proc, program, payload, nil, nil)
+}
+
+func (l *Libvirt) requestStream(proc uint32, program uint32, payload []byte, outStream io.Reader, inStream io.Writer) (response, error) {
 	serial := l.serial()
 	c := make(chan response)
 
 	l.register(serial, c)
+	defer l.deregister(serial)
+
+	err := l.sendPacket(serial, proc, program, payload, Call, StatusOK)
+	if err != nil {
+		return response{}, err
+	}
+
+	resp, err := l.getResponse(c)
+	if err != nil {
+		return resp, err
+	}
+
+	var abortOutStream chan bool
+	var outStreamErr chan error
+	if outStream != nil {
```

Comment:
the complexity of this block is too high

Reply:
this could still benefit from simplification

Reply:
I'm not sure I see how this can be made simpler... Could you please elaborate?

Reply:
The main problem is too many ifs make the code harder to reason about than it could be. A couple of simple changes that would help: 

in the for loop, change the second if to:
`if resp.Status != StatusContinue { 
    break 
}
if inStream != nil ....`

Define a function pointer called `abortOutStream()` above:

`abortOutStream := func() {}`

Then inside this block where you know outStream != nil, do this:

`abortOutStream = func() {abortOutStream <- true}`

That lets you remove two if statements, since it's always safe to call that function.

You could eliminate the last if too, by defering the logic it contains from here, but that's not as clear an improvement. Up to you.

Reply:
I'm sorry, I see now how I should've done this. I hope, new version will look simpler and easier for you too.